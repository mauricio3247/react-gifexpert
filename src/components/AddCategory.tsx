import React, { useState } from 'react'
import PropTypes from 'prop-types'
type CategoryProperties = {
    setCategories:  React.Dispatch<React.SetStateAction<string[]>>
}

export const AddCategory = ({setCategories}: CategoryProperties) => {

    const [inputValue, setInputValue] = useState('')
    const handleInput = (e:any) => {
        setInputValue(e.target.value)
    }

    const handleSubmit = (e: any) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories((categories)=> [inputValue, ...categories])
            setInputValue('')
        }
        
    }
    return (
        <form onSubmit= {handleSubmit}>
            <input 
            onChange={handleInput} 
            value={inputValue} />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}