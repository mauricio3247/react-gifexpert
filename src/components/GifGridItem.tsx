import React from 'react'
import PropTypes from 'prop-types'
export const GifGridItemComponent = ({title, url, id}:any) => {
    return (
        <div className="card animate__animated animate__zoomIn">
            <img src={url} alt={title} ></img>
            <p>{title}</p>
        </div>
    )
}

GifGridItemComponent.propTypes = {
    title: PropTypes.string.isRequired, 
    url: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired
}