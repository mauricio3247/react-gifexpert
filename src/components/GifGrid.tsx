import React from 'react'
import PropTypes from 'prop-types'
import { GifGridItemComponent } from './GifGridItem'

import { useFetchGifs } from '../hooks/useFetchGifs'

type GifGridProperties = {
    category: string
}

export const GifGridComponent = ({category}: GifGridProperties)=> {
    
    const {data:images, loading} = useFetchGifs(category);

    return (
        <>
        <h3 className="animate__animated animate__fadeIn">{category}</h3>
        <div className='card-grid'>
            
            { loading && <p className="animate__animated animate__flash">'Loading...'</p> }

            {
                images.map ( img => (
                    <GifGridItemComponent 
                    {...img}
                    key={img.id} />
                ))
            }
                

        </div>
        </>
    )
}

GifGridComponent.propTypes = {
    category: PropTypes.string.isRequired
}