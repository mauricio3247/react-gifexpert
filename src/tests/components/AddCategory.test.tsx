import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { AddCategory } from '../../components/AddCategory'

describe ('Pruebas en componente AddCategory', ()=> {

    const setCategories = jest.fn(); //permite crear un mock
    let wrapper: ShallowWrapper

    beforeEach(()=> {
        jest.clearAllMocks()
        wrapper = shallow(<AddCategory setCategories={setCategories} />)
    })


    test ('debe mostrarse el componente', ()=> {
        expect(wrapper).toMatchSnapshot()
    })

    test ('Debe de cambiar la caja de texto', ()=> {
        const value = 'Hola mundo'
        wrapper.find('input').simulate('change', {target: { value }});

        expect(wrapper.find('input').prop('value')).toBe(value)
    })

    test ('no debe postear la informacion en onSubmit', ()=> {
        wrapper.find('form').simulate('submit', {
            preventDefault: ()=> {}
        })
        expect(setCategories).not.toHaveBeenCalled()
    })

    test ('Debe llamar el setCategories y limpiar la caja de texto', () => {
        const value = 'Hola mundo'
        wrapper.find('input').simulate('change', {target: { value }});

        wrapper.find('form').simulate('submit', {
            preventDefault: ()=> {}
        })

        expect(setCategories).toHaveBeenCalled()
        /**
         * Espera que cuando la funcion sea llamada el parametro enviado sea un funcion como 
         * lo describe el componente
         */
        expect(setCategories).toBeCalledWith(expect.any(Function))
        expect(wrapper.find('input').prop('value')).toBe('')
    })
})