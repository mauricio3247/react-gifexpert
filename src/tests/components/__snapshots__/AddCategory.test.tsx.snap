// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente AddCategory debe mostrarse el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    value=""
  />
</form>
`;
